### DISCUSSION TITLE
Large applications structure

### DISCUSSION THREAD

**Discussion by lttr** [nuxt/nuxt#23773] [General]
During NuxtNation 2023 there was quite a few questions asked about layers, application structure, DDD etc.

I'm looking for some guidance. Still, after implemented several large sites in the last couple of years.

Let me first describe what am I looking for and how a typical project I work on look like: Imagine a set of websites or webapps that are very similar, but also different enough that it is not practical to only use configuration to differentiate them. On the other side the internals of those apps are quite complex. They have several features and each feature needs internationalization, data access layer, maybe authentication, documentation... All of the apps and features share some aspects - global ui components, eslint config, postcss config...

I think this is a well known problem space. For example monorepo tools like [NX](https://nx.dev/concepts/more-concepts/applications-and-libraries) or [Turborepo](https://turbo.build/repo/docs/handbook/workspaces) have a concept of `apps` and `libs` which correspond to what I have described above. The tricky part is how to structure all those libs. NX has some [ideas](https://nx.dev/concepts/more-concepts/creating-libraries) about libraries like when to create them, how to name them, which type of lib should not be imported by other type etc.

So what is challenging about this is how to actually implement this with Nuxt. Since we are talking about modular architecture - Nuxt modules come to mind. And while it looks like it is possible to do it (I am currently using Nuxt modules for everything in a monorepo with my team), it is not really straight-forward to do it. I will be using word lib for a building block or a as a generic word module.

What I'm missing specifically:
- A piece of documentation which would discuss this use-case and possible approaches
- A detailed comparison of possible implementation of libs: Nuxt layers vs Nuxt modules vs structure inside monolitic Nuxt app vs packages of Vue components vs plain JavaScript modules/packages...
- A tool to summarize information about what every lib is exposing
- What to look for in order to make lib dependencies (another libs) work for an app (I was not able to make this work for all aspects of an application with Nuxt layers, therefore we have been using Nuxt modules)
- How to use specific tools and libs to implement Hexagonal architecture or use DDD principles
- A template for such lib that has the kind of library like I have described above in mind (the template for Nuxt module that can be used with nuxi is more for a standalone module that is prepared to be released on npm)
- Is it possible to mix Nuxt layers, Nuxt modules, Vue packages etc. as libs?
- A way to declare which libs can import which libs (possible with NX)
- A way to visualize dependencies (possible with NX but only on the package manager level)

A couple of reasons why we should even care:
- When your app is modularized by feature you can better reason about each feature, how to maintain it, or how to remove it without breaking everything
- Having small modules supports reusability and flexibility (some modules might be released publicly, used by different projects)
- The build might be way faster thanks to monorepo tools which are able to cache individual module build outputs

Other discussions around this topic:
- a talk Layers Unwrapped: A Fresh Approach to Nuxt App Architecture by Krutie Patel during NuxtNation 2023
- https://github.com/nuxt/nuxt/discussions/20380
- https://github.com/nuxt/nuxt/discussions/16373
- https://github.com/nuxt/nuxt/discussions/10071
- https://deltener.com/blog/nuxt-enterprise-patterns-component-management/

I would like to collaborate on this topic if anyone is interested...

**Comment by davestewart** [comment]
These may be of interest:

- Nuxt 3: [Modular site architecture with Nuxt layers](https://davestewart.co.uk/blog/nuxt-layers/) (additional resources listed at end of article)
- Nuxt 2: [Nuxt Areas](https://github.com/davestewart/nuxt-areas)
